src/frontend/query_commands.ml

```diff
@@ -481,30 +481,48 @@ let dispatch pipeline (type a) : a Query_protocol.t -> a =
     let typer = Mpipeline.typer_result pipeline in
     let pos = Mpipeline.get_lexing_pos pipeline pos in
     begin match Mbrowse.select_open_node (Mtyper.node_at typer pos) with
-      | None | Some (_, []) -> []
-      | Some (path, ((_, node) :: _)) ->
+      | None | Some (_, _, []) -> []
+      | Some (path, longident, ((_, node) :: _)) ->
         let paths =
           Browse_tree.all_occurrences_of_prefix ~strict_prefix:true path node in
         let paths = List.concat_map ~f:snd paths in
-        let rec path_to_string acc (p : Path.t) =
-          match p with
-          | Pident ident ->
-            String.concat ~sep:"." (Ident.name ident :: acc)
-          | Pdot (path', s) when
-              mode = `Unqualify && Path.same path path' ->
-            String.concat ~sep:"." (s :: acc)
-          | Pdot (path', s) ->
-            path_to_string (s :: acc) path'
-          | _ -> raise Not_found
+        let leftmost_ident = Longident.flatten longident |> List.hd in
+        let qual_or_unqual_path p =
+          let rec aux acc (p : Path.t) =
+            match p with
+            | Pident ident ->
+              Ident.name ident :: acc
+            | Pdot (path', s) when
+                mode = `Unqualify && Path.same path path' ->
+              s :: acc
+            | Pdot (path', s) when
+                mode = `Qualify && s = leftmost_ident ->
+              s :: acc
+            | Pdot (path', s) ->
+              aux (s :: acc) path'
+            | _ -> raise Not_found
+          in
+          aux [] p |> String.concat ~sep:"."
+        in
+        (* checks if the (un)qualified longident has a different length, i.e., has changed
+
+           XXX(Ulugbek): computes longident length using [loc_start] and [loc_end], hence
+           it doesn't work for multiline longidents because we can't compute their length *)
```

Comment:
Perhaps `qual_or_unqual_path` should just return the list, without concatenating the elements, and then you could just compare that to the result of `Longident.flatten`?

Reply:
Let me describe how I see this: `Browse_tree.all_occurrences_of_prefix` returns `(Browse_raw.node * Path.t Location.loc list) list`. This means that we get a node `n` and the absolute path `p` associated with it, but we do not know the identifier `i` of the node `n`. However, we need to check whether `[qual_or_unqual_path] p = i` to be able to tell whether the identifier is updated and we should return that result. (AFAIU, `Ident.t` represents the identifier used in the source code, while `Path.t` is the absolute path, which can be different from that identifier, e.g., `foo` vs `Dune__exe.Import.foo`)

If I understand it correctly, we need to either get the ident either from node `n` or make sure that `Browse_tree.all_occurrences_of_prefix` returns also idents along with paths, which is quite an invasive change. Does this make sense? If yes, which solution would you prefer me to work on? 

---

The motivation behind this PR is to have smarter code actions (code change proposal based on the current context) 

For example, in the screenshot below, we get a code action `Remove module name from the identifiers` because there are meaningful changes possible to the code (remove `Unix` from `Unix.putenv`). Without this PR, this code action is proposed even if it doesn't change the code. 

![image](https://user-images.githubusercontent.com/16353531/124577889-23e2dc80-de67-11eb-9059-a0d8393dd5c5.png)

Reply:
Sorry, my message wasn't clear enough. But I think I understood what you're trying to achieve alright.
My point was just that, as you say, comparing string lengths is brittle. And you're going to run into trouble in more cases than just multi-line longidents. Anytime you have a blank space in the source is going to be problematic for you.

So what I was suggesting is to not count the raw/textual length, but the number of elements of the path.
Something like this:
```ocaml
let qual_or_unqual_path p =
  let rec aux acc (p : Path.t) =
    match p with
    | Pident ident ->
      Ident.name ident :: acc
    | Pdot (path', s) when mode = `Unqualify && Path.same path path' ->
      s :: acc
    | Pdot (path', s) when mode = `Qualify && s = leftmost_ident ->
      s :: acc
    | Pdot (path', s) ->
      aux (s :: acc) path'
    | _ -> raise Not_found
  in
  aux [] p
in
let same_path old_path new_path =
  let rec path_len = function
    | Pident _ -> 1
    | Pdot (p, _) -> 1 + path_len p
    | _ -> assert false
  in
  path_len old_path = List.length old_path
in
...
    match qual_or_unqual_path path with
    | new_path when same_path path new_path -> None
    | new_path -> Some (String.concat ~sep:"." new_path, loc)
    | exception Not_found -> None
...
```

What's wrong with that?

Reply:
Given ident = `foo`, old_path = `Import.foo`, we call unqualify `Import` on the old_path and get new_path = `foo`, right? `same_path old_path new_path` returns false and we return this as an edit, even though the user already has ident as `foo` and doesn't need this edit. Does this make sense? 

(Again, I'm under assumption that `Path.t` we get from `Browse_tree.all_occurrences_of_prefix` represents the absolute path for the node, which can be different from the ident used for that node)

Reply:
Nevermind, I got confused. At first I thought you had the longident corresponding to the path, which would have been ideal. But you don't. Then I tried to adapt my solution to use the unedited path, but you're right that that one is always absolute, so that won't work either.

I think we can revisit this in the future, and that just playing with locations for now is fine.